File: src/main/kotlin/specscript/files/PackageRegistry.kt
package specscript.files

import specscript.language.canonicalCommandName
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.exists
import kotlin.io.path.isDirectory
import kotlin.io.path.name

typealias PackageName = String

object PackageRegistry {

    var packagePath: Path? = null
    var autoPackagePath: Path? = null

    fun findPackage(name: PackageName): Path? {
        for (libDir in searchPath()) {
            val candidate = libDir.resolve(name)
            if (candidate.isDirectory() && isPackage(candidate, name)) {
                return candidate
            }
        }
        return null
    }

    fun findEnclosingPackageLibrary(startDir: Path): Path? {
        var dir: Path? = startDir.toAbsolutePath().normalize()
        while (dir != null) {
            val info = SpecScriptDirectories.get(dir)
            if (info.packageInfo != null) {
                return dir.parent
            }
            dir = dir.parent
        }
        return null
    }

    private fun searchPath(): List<Path> {
        val paths = mutableListOf<Path>()

        autoPackagePath?.let { paths.add(it) }

        packagePath?.let { paths.add(it) }

        System.getenv("SPECSCRIPT_PACKAGE_PATH")?.let { envPath ->
            envPath.split(":").filter { it.isNotBlank() }.forEach {
                paths.add(Path.of(it))
            }
        }

        val defaultDir = Path.of(System.getProperty("user.home"), ".specscript", "packages")
        if (defaultDir.exists()) {
            paths.add(defaultDir)
        }

        return paths
    }

    private fun isPackage(dir: Path, expectedName: PackageName): Boolean {
        val info = SpecScriptDirectories.get(dir)
        return info.packageInfo != null
    }

    fun scanCommands(
        packageDir: Path,
        items: List<ImportItem>
    ): Map<String, SpecScriptFile> {
        val commands = mutableMapOf<String, SpecScriptFile>()

        for (item in items) {
            when (item) {
                is ImportItem.Command -> {
                    val file = packageDir.resolve("${item.path}$YAML_SPEC_EXTENSION")
                    if (file.exists()) {
                        val name = item.alias
                            ?: canonicalCommandName(asScriptCommand(file.name))
                        commands[canonicalCommandName(name)] = SpecScriptFile(file)
                    }
                }

                is ImportItem.Name -> {
                    addNameImport(commands, packageDir, item.value, item.alias)
                }

                is ImportItem.Directory -> {
                    val dir = packageDir.resolve(item.path)
                    if (dir.isDirectory()) {
                        addDirectoryCommands(commands, dir)
                    }
                }

                is ImportItem.Wildcard -> {
                    val dir = if (item.path.isEmpty()) packageDir
                    else packageDir.resolve(item.path)

                    if (dir.isDirectory()) {
                        if (item.recursive) {
                            addRecursiveCommands(commands, dir)
                        }

                        else {
                            addDirectoryCommands(commands, dir)
                        }
                    }
                }

            }
        }

        return commands
    }

    fun scanLocalCommands(
        configDir: Path,
        localPath: String,
        items: List<ImportItem>
    ): Map<String, SpecScriptFile> {
        val resolvedPath = localPath.removePrefix("./")
        val dir = configDir.resolve(resolvedPath)
        val commands = mutableMapOf<String, SpecScriptFile>()

        for (item in items) {
            when (item) {
                is ImportItem.Command -> {
                    val file = dir.resolve("${item.name}$YAML_SPEC_EXTENSION")
                    if (file.exists()) {
                        val name = item.alias
                            ?: canonicalCommandName(asScriptCommand(file.name))
                        commands[canonicalCommandName(name)] = SpecScriptFile(file)
                    }
                }

                is ImportItem.Name -> {
                    addNameImport(commands, dir, item.value, item.alias)
                }

                is ImportItem.Wildcard -> {
                    if (item.recursive) {
                        addRecursiveCommands(commands, dir)
                    }

                    else {
                        addDirectoryCommands(commands, dir)
                    }
                }

                is ImportItem.Directory -> {
                    val subDir = dir.resolve(item.path)
                    if (subDir.isDirectory()) {
                        addDirectoryCommands(commands, subDir)
                    }
                }

            }
        }

        return commands
    }

    private fun addNameImport(
        commands: MutableMap<String, SpecScriptFile>,
        baseDir: Path,
        name: String,
        alias: String? = null
    ) {
        val asFile = baseDir.resolve("$name$YAML_SPEC_EXTENSION")
        if (asFile.exists()) {
            val commandName = alias
                ?: canonicalCommandName(asScriptCommand(asFile.name))
            commands[canonicalCommandName(commandName)] = SpecScriptFile(asFile)
            return
        }

        val asDir = baseDir.resolve(name)
        if (asDir.isDirectory()) {
            addDirectoryCommands(commands, asDir)
        }
    }

    private fun addDirectoryCommands(
        commands: MutableMap<String, SpecScriptFile>,
        dir: Path
    ) {
        if (!dir.isDirectory()) return

        Files.list(dir).use { stream ->
            stream.filter { it.name.endsWith(YAML_SPEC_EXTENSION) || it.name.endsWith(MARKDOWN_SPEC_EXTENSION) }
                .filter { !it.isDirectory() }
                .forEach { file ->
                    val commandName = asScriptCommand(file.name)
                    commands[canonicalCommandName(commandName)] = SpecScriptFile(file)
                }
        }
    }

    private fun addRecursiveCommands(commands: MutableMap<String, SpecScriptFile>, dir: Path) {
        if (!dir.isDirectory()) return

        addDirectoryCommands(commands, dir)

        Files.list(dir).use { stream ->
            stream.filter { it.isDirectory() && !isExcluded(it) }
                .forEach { subDir ->
                    addRecursiveCommands(commands, subDir)
                }
        }
    }

    private fun isExcluded(dir: Path): Boolean {
        if (dir.name == "tests") return true

        val info = SpecScriptDirectories.get(dir)
        return info.hidden
    }
}

File: typescript/src/language/package-registry.ts
import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs'
import { join, resolve, dirname } from 'node:path'
import { homedir } from 'node:os'
import { parseYaml } from '../util/yaml.js'
import { canonicalName, type CommandHandler } from './command-handler.js'
import { createFileCommandHandler, fileToCommandName } from './context.js'
import type { ImportItem } from './package-import.js'
import type { JsonObject } from './types.js'

const YAML_SPEC_EXTENSION = '.spec.yaml'
const MARKDOWN_SPEC_EXTENSION = '.spec.md'

export let packagePath: string | undefined
export let autoPackagePath: string | undefined

export function setPackagePath(path: string | undefined): void {
  packagePath = path
}

export function setAutoPackagePath(path: string | undefined): void {
  autoPackagePath = path
}

export function findEnclosingPackageLibrary(startDir: string): string | undefined {
  let dir: string | undefined = resolve(startDir)
  while (dir) {
    const config = readConfig(dir)
    if (config && 'Package info' in config) {
      return dirname(dir)
    }
    const parent = dirname(dir)
    if (parent === dir) break
    dir = parent
  }
  return undefined
}

export function findPackage(name: string): string | undefined {
  for (const libDir of searchPath()) {
    const candidate = join(libDir, name)
    try {
      if (statSync(candidate).isDirectory() && isPackage(candidate, name)) {
        return candidate
      }
    } catch {
      // Directory not accessible
    }
  }
  return undefined
}

function searchPath(): string[] {
  const paths: string[] = []

  if (autoPackagePath) {
    paths.push(autoPackagePath)
  }

  if (packagePath) {
    paths.push(packagePath)
  }

  const envPath = process.env['SPECSCRIPT_PACKAGE_PATH']
  if (envPath) {
    for (const p of envPath.split(':')) {
      if (p.trim()) paths.push(p.trim())
    }
  }

  const defaultDir = join(homedir(), '.specscript', 'packages')
  if (existsSync(defaultDir)) {
    paths.push(defaultDir)
  }

  return paths
}

function isPackage(dir: string, expectedName: string): boolean {
  const config = readConfig(dir)
  if (!config) return false
  return 'Package info' in config
}

function readConfig(dir: string): JsonObject | undefined {
  const configPath = join(dir, 'specscript-config.yaml')
  try {
    const content = readFileSync(configPath, 'utf-8')
    const parsed = parseYaml(content)
    if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) {
      return parsed as JsonObject
    }
  } catch {
    // No config or not readable
  }
  return undefined
}

export function scanCommands(
  packageDir: string,
  items: ImportItem[],
): Map<string, CommandHandler> {
  const commands = new Map<string, CommandHandler>()

  for (const item of items) {
    switch (item.kind) {
      case 'command': {
        const file = join(packageDir, `${item.path}${YAML_SPEC_EXTENSION}`)
        if (existsSync(file)) {
          const baseName = fileToCommandName(`${item.name}${YAML_SPEC_EXTENSION}`)
          const name = item.alias ?? baseName
          if (name) commands.set(canonicalName(name), createFileCommandHandler(name, file))
        }
        break
      }

      case 'name':
        addNameImport(commands, packageDir, item.value, item.alias)
        break

      case 'directory': {
        const dir = join(packageDir, item.path)
        if (isDirectory(dir)) {
          addDirectoryCommands(commands, dir)
        }
        break
      }

      case 'wildcard': {
        const dir = item.path ? join(packageDir, item.path) : packageDir
        if (isDirectory(dir)) {
          if (item.recursive) {
            addRecursiveCommands(commands, dir)
          } else {
            addDirectoryCommands(commands, dir)
          }
        }
        break
      }
    }
  }

  return commands
}

export function scanLocalCommands(
  configDir: string,
  localPath: string,
  items: ImportItem[],
): Map<string, CommandHandler> {
  const resolvedPath = localPath.replace(/^\.\//, '')
  const dir = resolve(configDir, resolvedPath)
  const commands = new Map<string, CommandHandler>()

  for (const item of items) {
    switch (item.kind) {
      case 'command': {
        const file = join(dir, `${item.name}${YAML_SPEC_EXTENSION}`)
        if (existsSync(file)) {
          const baseName = fileToCommandName(`${item.name}${YAML_SPEC_EXTENSION}`)
          const name = item.alias ?? baseName
          if (name) commands.set(canonicalName(name), createFileCommandHandler(name, file))
        }
        break
      }

      case 'name':
        addNameImport(commands, dir, item.value, item.alias)
        break

      case 'wildcard':
        if (item.recursive) {
          addRecursiveCommands(commands, dir)
        } else {
          addDirectoryCommands(commands, dir)
        }
        break

      case 'directory': {
        const subDir = join(dir, item.path)
        if (isDirectory(subDir)) {
          addDirectoryCommands(commands, subDir)
        }
        break
      }
    }
  }

  return commands
}

function addNameImport(
  commands: Map<string, CommandHandler>,
  baseDir: string,
  name: string,
  alias?: string,
): void {
  const asFile = join(baseDir, `${name}${YAML_SPEC_EXTENSION}`)
  if (existsSync(asFile)) {
    const baseName = fileToCommandName(`${name}${YAML_SPEC_EXTENSION}`)
    const commandName = alias ?? baseName
    if (commandName) {
      commands.set(canonicalName(commandName), createFileCommandHandler(commandName, asFile))
    }
    return
  }

  const asDir = join(baseDir, name)
  if (isDirectory(asDir)) {
    addDirectoryCommands(commands, asDir)
  }
}

function addDirectoryCommands(
  commands: Map<string, CommandHandler>,
  dir: string,
): void {
  if (!isDirectory(dir)) return

  try {
    for (const entry of readdirSync(dir)) {
      if (!entry.endsWith(YAML_SPEC_EXTENSION) && !entry.endsWith(MARKDOWN_SPEC_EXTENSION)) continue
      const filePath = join(dir, entry)
      if (statSync(filePath).isDirectory()) continue

      const commandName = fileToCommandName(entry)
      if (!commandName) continue
      commands.set(canonicalName(commandName), createFileCommandHandler(commandName, filePath))
    }
  } catch {
    // Directory not readable
  }
}

function addRecursiveCommands(commands: Map<string, CommandHandler>, dir: string): void {
  if (!isDirectory(dir)) return

  addDirectoryCommands(commands, dir, undefined)

  try {
    for (const entry of readdirSync(dir)) {
      const fullPath = join(dir, entry)
      if (isDirectory(fullPath) && !isExcluded(entry)) {
        addRecursiveCommands(commands, fullPath)
      }
    }
  } catch {
    // Directory not readable
  }
}

function isExcluded(dirName: string): boolean {
  if (dirName === 'tests') return true
  // Check for hidden flag in config
  const config = readConfig(dirName)
  return config?.['hidden'] === true
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory()
  } catch {
    return false
  }
}
